doc/md/concepts: migration directory integrity (#1083)
ariga/atlas: doc/md/concepts/migration-directory-integrity.md

---
id: migration-directory-integrity
slug: /concepts/migration-directory-integrity
title: Migration Directory Integrity File
---

Applying changes to a database schema in the wrong order can be dangerous. For this
reason, Atlas is built on a workflow that enforces a _linear history_ using a 
_migration directory integrity file_.

### The problem

> Suppose you have multiple teams working on a system in parallel, all of which need to make 

Reviewer: ```suggestion
Suppose you have multiple teams working on a system simultaneously, all of which need to make 
```

changes to the database schema from time to time. Unless they somehow coordinate, they
may end up with a broken migration directory.

Consider what would happen if Team A and B, working in parallel, both merge a migration
that creates a new table named `inventory`. This is illustrated in this diagram:

![Teams committing migrations in parallel. Source: Ent blog](https://entgo.io/images/assets/migrate/no-conflict.svg)

Something like this might happen:
1. Team A creates a feature branch, committing a migration creating the `inventory` table. 
2. Team B creates a second feature branch, also creating a table by the same name. 
3. Both branches pass code-review and continuous integration. 
4. Team A's branch is merged to the mainline branch. 
5. Team B's branch is merged. 
6. When both changes are deployed, the first one to run will succeed and the second will fail.

>    This will happen in an arbitrary order. (Migrations are run in lexicographic order, usually set by the timestamp

Reviewer: ```suggestion
   This will happen in an arbitrary order (migrations are run in lexicographic order, usually set by the timestamp
```

   on the developer's workstation when generating them).

> 

Reviewer 2: There even is another problem. Consider:

Team A generatee migration (version 1)
Team B generates migration (version 2)
Team B merges and the migration is executed (now version 2 is deployed)
Team A merges second with migration version 1. Depending on the tool, version 1 will never get applied since currently the database is already at version 2. Or it gets merged (and might succeed). But the migration order for this depliyment was 2->1. 

Somtimes in the futrue a new deployment gets created. Now the migration order is 1->2. This is a problem as well the file guards against.

Author: Cool
Maybe we can add another section outlining this to prevent something too lengthy

### The solution

Recovering from a failed migration is quite a headache, so wouldn't it be great
to prevent this from ever happening?

Code conflicts are usually detected by source-control systems (such as Git) when
the same line in the same file is modified by two different commits. In our case,
no such conflict happens because migrations are typically described in a separate
file for each migration. 

> Atlas's engine offers a way to prevent concurrent creation of new migration 

Reviewer: ```suggestion
Atlas's engine offers a way to prevent concurrent creations of new migration 
```

> files and guard against accidental changes in the migration history we call

Reviewer: ```suggestion
files and guards against accidental changes in the migration history we call
```

> *Migration Directory Integrity File*, which simply is another file in your

Reviewer: ```suggestion
*Migration Directory Integrity File*. This file is simply another file in your
```

> migration directory called `atlas.sum` and looks something like:

Reviewer: ```suggestion
migration directory called `atlas.sum` and looks something like:
```

```text
h1:KRFsSi68ZOarsQAJZ1mfSiMSkIOZlMq4RzyF//Pwf8A=
20220318104614_team_A.sql h1:EGknG5Y6GQYrc4W8e/r3S61Aqx2p+NmQyVz/2m8ZNwA=
```

The `atlas.sum` file contains the checksum of each migration file (implemented by a 
reverse, one branch merkle hash tree), and a sum of all files. Adding new files 
results in a change to the sum file, which will raise merge conflicts in most 

> version controls systems. 

Reviewer: ```suggestion
version control systems. 
```
